### PR TITLE
Bug Fix - Vector3 project on plane

### PR DESCRIPTION
**Pull request by strutcode**
This proposal fixes an apparent incorrect calculation in the `Vector3.projectOnPlaneToRef` method which returns confusing results for many planes, especially those not centered on the world origin.

A couple of unit tests were added to logically demonstrate the failure of the existing code, as well as a new original implementation which seems to work correctly for various cases.

Playground demonstrating the behavior for testing:
https://playground.babylonjs.com/#MHMIPM

### CODE REVIEW

File: packages/dev/core/src/Maths/math.vector.ts [hunk 1]
```diff
@@ -1150,22 +1150,43 @@ export class Vector3 {
...
+
+        // Since the normal in Babylon should point toward the viewer, invert it for the dot product
+        const inverseNormal = MathTmp.Vector3[1];
+        plane.normal.negateToRef(inverseNormal);
+        
+        // This vector is the direction
+        const direction = this;
+
+        // Set a margin to prevent issues with directions very close to perpendicular to the plane
+        const epsilon = 0.000001;
```
Comment (RaananW) [inline packages/dev/core/src/Maths/math.vector.ts:1165]: This exists in math.constants (and is already imported to the math module)

File: packages/dev/core/src/Maths/math.vector.ts [hunk 2]
```diff
@@ -1150,22 +1150,43 @@ export class Vector3 {
...
+        // Use the normal scaled to the plane offset as the origin for the formula
+        const planeOrigin = MathTmp.Vector3[0];
+        plane.normal.scaleToRef(-plane.d, planeOrigin)
+
+        // Since the normal in Babylon should point toward the viewer, invert it for the dot product
+        const inverseNormal = MathTmp.Vector3[1];
+        plane.normal.negateToRef(inverseNormal);
+        
+        // This vector is the direction
+        const direction = this;
```
Comment (RaananW) [inline packages/dev/core/src/Maths/math.vector.ts:1162]: why is that needed?

**Review by RaananW — COMMENTED** [review]
LGTM!

this is your playground with the snapshot - https://playground.babylonjs.com/?snapshot=refs/pull/12663/merge#MHMIPM

Just two questions about the code itself, and you have a formatting issue. you can run `npm run format:fix` if you don't use a formatter directly in your IDE.

File: packages/dev/core/src/Maths/math.vector.ts (hunk 1 above)
Comment (strutcode) [reply packages/dev/core/src/Maths/math.vector.ts:1165]: Good call! It's probably better to use a consistent fudge value. :)

File: packages/dev/core/src/Maths/math.vector.ts (hunk 2 above)
Comment (strutcode) [reply packages/dev/core/src/Maths/math.vector.ts:1162]: There's no reason, it's just a habit of mine to name things in a way that makes it as clear as possible what the code is doing. I believe it should be optimized out so it's purely aesthetic.

**Comment by strutcode** [comment]
Strange! I already fixed the formatting issues. I wonder why it's not showing up. I'll figure it out though.

File: packages/dev/core/src/Maths/math.vector.ts (hunk 2 above)
Comment (strutcode) [reply packages/dev/core/src/Maths/math.vector.ts:1162]: Let me know if I should change this or if it's good as is. :+1:

**Review by RaananW — APPROVED** [review]

**Comment by sebavan** [comment]
@strutcode would be great to for you to have a look at https://forum.babylonjs.com/t/project-3d-point-on-plane/9862/18?u=sebavan and maybe create a separate function for your behavior ?
